keps/0008-node-heartbeat.md

```diff
@@ -0,0 +1,374 @@
+- Kubelet sending heartbeats every 10s
+
+This proposal presents a proper solution for that problem.
+
+
+Note that currently:
+
+- Lack of NodeStatus update for 40s results in NodeController marking node
+  as NotReady (pods are no longer scheduled on that node)
+- Lack of NodeStatus updates for 5m results in NodeController starting pod
+  evictions from that node
+
+We would like to preserve that behavior.
+
+
+### Goals
+
+- Reduce size of etcd by making node heartbeats cheaper
+
+### Non-Goals
+
+The following are nice-to-haves, but not primary goals:
+
+- Reduce resource usage (cpu/memory) of control plane (e.g. due to processing
+  less and/or smaller objects)
+- Reduce watch-related load on Node objects
+
+## Proposal
+
+We propose introducing a new `LeaderElection` CRD with the following schema:
+```
+type LeaderElection struct {
+  metav1.TypeMeta `json:",inline"`
+  // Standard object's metadata.
+  // More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+  // +optional
+	ObjectMeta metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  // Specification of the LeaderElection.
+  // More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+  // +optional
+  Spec LeaderElectionSpec `json:"spec,omitempty"`
+}
+
+type LeaderElectionSpec struct {
+  HolderIdentity       string      `json:"holderIdentity"`
+  LeaseDurationSeconds int32       `json:"leaseDurationSeconds"`
+  AcquireTime          metav1.Time `json:"acquireTime"`
+  RenewTime            metav1.Time `json:"renewTime"`
+  LeaderTransitions    int32       `json:"leaderTransitions"`
+}
+```
+
+The Spec is effectively of already existing (and thus proved) [LeaderElectionRecord][].
+That would hopefully allow us go get directly to Beta.
+
+We will use CRD (as opposed to built-in API), because:
+
+- CRDs are `the` way to created new APIs
+- even though CRDs aren't very efficient due to lack of protobuf support now:
+    - performance should be acceptable (because LeaderElection object will be small)
+      <br> TODO(wojtek-t): Working on microbenchmark to prove with data.
+    - protobuf support for CRDs is on the roadmap (though not near term)
+
+We will change Kubelet so that:
+
+1. Kubelet is periodically computing NodeStatus every 10s (at it is now), but that will
+   be independent from reporting status
+1. Kubelet is reporting NodeStatus if:
+   - there was a meaningful change in it (initially we can probably assume that every
+     change is meaningful, including e.g. images on the node)
+   - or it didn’t report it over last `node-status-update-period` seconds
+1. Kubelet creates and periodically updates its own LeaderElection object and frequency
+   of those updates is independent from NodeStatus update frequency.
+
+In the meantime, we will change `NodeController` to treat both updates of NodeStatus
+object as well as updates of the new `LeaderElection` object corresponding to a given
```

Comment:
This breaks backwards compatibility for any integration anyone has done to watching node objects for changes.  That’s any custom scheduler, and alerting based function, and probably a non trivial amount of custom node integrations built by users.

Reply:
Treating both as a healthiness signal? Why? Can you clarify that?

Reply:
We'll still write the status, just less frequently. It's effectively a configuration change.

Reply:
Also, schedulers shouldn't be watching elapsed time since last heartbeat directly. The node controller encodes cordoning and eviction in taints and conditions.

Reply:
@smarterclayton We thought about such compatibility issues and this is why we still updated lastHeartbeatTime periodically, but with a increasing period.